booster/library/Booster/Pattern/Rewrite.hs

```diff
@@ -801,108 +792,113 @@ performRewrite doTracing def mLlvmLibrary smtSolver varsToAvoid mbMaxDepth cutLa
+                                    _other -> withPatternContext pat' $ error "simplifyResult: Unexpected return value"
+                            Right (cutPoint@(RewriteCutPoint lbl _ _ _), _) -> withPatternContext pat' $ do
+                                simplified <- simplifyResult pat' cutPoint
+                                case simplified of
+                                    RewriteCutPoint{} ->
+                                        logMessage $ "Cut point " <> lbl <> " after " <> showCounter counter
+                                    RewriteStuck{} ->
+                                        logMessage $ "Stuck after " <> showCounter counter
+                                    RewriteTrivial{} ->
+                                        logMessage $ "Simplified to bottom after " <> showCounter counter
+                                    _other -> error "simplifyResult: Unexpected return value"
+                                pure simplified
+                            Right (stuck@RewriteStuck{}, cache) -> do
+                                logMessage $ "Stopped after " <> showCounter counter
+                                updateCache cache
+                                emitRewriteTrace $ RewriteStepFailed $ NoApplicableRules pat'
+                                if wasSimplified
+                                    then pure stuck
+                                    else withSimplified pat' "Retrying with simplified pattern" (doSteps True)
+                            Right (trivial@RewriteTrivial{}, _) -> withPatternContext pat' $ do
+                                logMessage $ "Simplified to bottom after " <> showCounter counter
+                                pure trivial
+                            Right (aborted@RewriteAborted{}, _) -> withPatternContext pat' $ do
                                 logMessage $ "Aborted after " <> showCounter counter
-                                pure (RewriteAborted failure pat')
-                        Left failure -> do
-                            emitRewriteTrace $ RewriteStepFailed failure
-                            let msg = "Aborted after " <> showCounter counter
-                            if wasSimplified
-                                then logMessage msg >> pure (RewriteAborted failure pat')
-                                else withSimplified pat' msg (pure . RewriteAborted failure)
+                                simplifyResult pat' aborted
+                            -- if unification was unclear and the pattern was
+                            -- unsimplified, simplify and retry rewriting once
+                            Left failure@(RuleApplicationUnclear rule _ remainder)
+                                | not wasSimplified -> do
+                                    emitRewriteTrace $ RewriteStepFailed failure
+                                    -- simplify remainders, substitute and rerun.
```

Comment:
I've just noticed that this comment is irrelevant --- not sure how it got here.

Reply:
That comment is old: https://github.com/runtimeverification/haskell-backend/pull/4010/files/cd2a4dd8a890cead5e7612713422ab7e93e9fc01#diff-79733ce60c52de929f411c95349bc4670bf6aa92709665c7bf313455787b9bbeL880-L881
I think it comes from an abandoned experiment with matching remainders that was backed out some time last year.